WIP: Add support for Arturo language

RickBarretto opened the pull request:
This PR adds Arturo Language highlight support. 🙂

> _Simple, expressive & portable programming language for efficient scripting_

[Arturo's Github](https://github.com/arturo-lang/arturo)[
Arturo's Website](http://arturo-lang.io/)

RickBarretto commented:
Sorry, I tried to use this function. (It was running on my local machine. Generating the files correctly).
But, I replaced for manual configuration, since it's breaking in CI tests.

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+        self.handle_annotateds = get_bool_opt(options, 'handle_annotateds', True)
+        RegexLexer.__init__(self, **options)
+
+__all__ = [
```
Please put this near the top of the file, after the imports.

RickBarretto replied:
[d80b451](https://github.com/pygments/pygments/pull/2259/commits/d80b451d702275f94707a12044feb5f22568e1de), done!

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+            (r'(;)(.*)$', Comment.Single),
+            include('shebang'),
+        ],
+            'shebang': [
```
I see you've used additional indentation as a kind of hierarchy marker here, please don't (it messes with style checking)

RickBarretto replied:
- [x] [remove additional indentation](https://github.com/pygments/pygments/pull/2259/commits/a5a0f89bf66415703f600b8895ca3c7191b4575e)

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+            include('constants'),
+            include('builtin_functions'),
+            include('operators'),
+            (r'.', Name),
```
This seems odd - anything can be a Name?

Uncategorized "misc" should be `Text`.

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+        lexer = None
+        if self.handle_annotateds:
+            try:
+                lexer = get_lexer_by_name( match.group(3).strip() )
```
```suggestion
                lexer = get_lexer_by_name(match.group(3).strip())
```

RickBarretto replied:
[remove whitespace](https://github.com/pygments/pygments/pull/2259/commits/31dc1cd6eed6706a6d367e529097e9eca5fc39be)

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+import re
```
standard file header is missing...

RickBarretto replied:
- [x] [add file header](https://github.com/pygments/pygments/pull/2259/commits/a0ff99bc43bd98f1dd96f43afffe6cadf9d08829)

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+class ArturoLexer(RegexLexer):
+
+    name = 'Arturo'
+    aliases = ['Arturo', 'Art', 'arturo', 'art']
```
aliases should be only lowercase.

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+        ],
+
+        'comments': [
+            (r'(;)(.*)$', Comment.Single),
```
no need to make `()` groups when you're not using them for `bygroups` or alternates - same applies for other rules below

also, `.*` should be `.*?` here

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+            include('shebang'),
+        ],
+            'shebang': [
+                (r'^((\s#!)|(#!))(.*)$',
```
`.*` should be `.*?` here.

RickBarretto replied:
- [x] [no need to group it](https://github.com/pygments/pygments/pull/2259/commits/c7987f2b004170cb0a8944e68f0f1fda594a8e05)
- [x] [add lazy quantifier](https://github.com/pygments/pygments/pull/2259/commits/e3fc6fb194354d09524799b0ce3b7d86f1fc84bd)

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+            include('punctuation'),
+            include('sugar'),
+
+            (r'<:|:>|:<|:>'
```
better use `words()` for such lists of literals, it is easier to read and avoids common mistakes - for example, this will never match `!!` since the `!` alternate comes first.

RickBarretto replied:
In this case, do I should escape special characters as `?`?

RickBarretto replied:
- [x] [replace with words function](https://github.com/pygments/pygments/pull/2259/commits/485b840311819421b1744f630f9490baaa53dda6)

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+            ],
+
+        'constants': [
+            include('boolean'),
```
please don't create new states for these, they are only used once.

RickBarretto replied:
> * [x]  [remove additional state](https://github.com/pygments/pygments/pull/2259/commits/ddb7923d1aa888f2bce3577bc44f744ad4f55c32)
>   
>   * I tried to just remove `constants` state, but I think the commit bellow is better
> * [x]  [remove the high amount of trivial states](https://github.com/pygments/pygments/pull/2259/commits/cb6a6b7595b15e8971794adc701e8eea1730198b)
>   
>   * I just removed that amount of states, put on `constants`, and added side notes. I think it's better.

https://github.com/pygments/pygments/pull/2259#discussion_r1004408549

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+                ],
+                    'inside-interpol': [
+                        (r'\|', String.Interpol, '#pop'),
+                        (r'.', using(this)),
```
this won't work propery, since it will pass single characters to the `this` instance.

rule should be `r'[^|]+'`

RickBarretto replied:
Wow, that explains a lot to me!

RickBarretto replied:
- [x] [fix interpolation rule](https://github.com/pygments/pygments/pull/2259/commits/2c08d2242c2145909ebd602a76dee0a8955bf33a)

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+                    'inside-template': [
+                        (r'\|\|\>',
+                            String.Interpol, '#pop'),
+                        (r'.', using(this)),
```
see above

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+                    (r'\\"',  String.Escape), # Escaping Quote Character
+                ],
+            'string-content-single-line': [
+                (r'.', String)
```
again this is bad since it matches the string contents char-by-char

You can see it in the test golden output file

RickBarretto replied:
- [x] [replace inefficient match](https://github.com/pygments/pygments/pull/2259/commits/7b683c48e21bb507e78ff8a549a4211455541c1d)

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+        ],
+            'builtin-predicate-functions': [
+                (
+                    r'\b('
```
use words() again...

RickBarretto replied:
In this case, `words` aren't working to me.
It's the reason I used this handmade regex.

RickBarretto replied:
- [x] [replace with words function](https://github.com/pygments/pygments/pull/2259/commits/485b840311819421b1744f630f9490baaa53dda6)

RickBarretto replied:
- [x] [replace hand-made regex with words function](https://github.com/pygments/pygments/pull/2259/commits/2a5b4d602f07a9bd71910e73134e707b8bcb5a64)

I done! But I got a question. 
When I use it, this doesn't match the `?` at the end on syntax highlight. Look:
![image](https://user-images.githubusercontent.com/78623871/197778278-a043adb0-8666-4818-9a44-2524b8055b0f.png)

As You can see, `or` is matched as `Name.Builtin`, but, `?` mark - that is part of `or` - doesn't. 
So, `or?` should be matched as a `Name.Builtin`, not just `or`.

Just as labels are:
![image](https://user-images.githubusercontent.com/78623871/197778991-15b610bb-a759-4c2f-ad59-12fd2c8e6427.png)

RickBarretto replied:
- [x] [add missing r string prefix](https://github.com/pygments/pygments/pull/2259/commits/aa87683cb1f4ec9bd3ed728d4d63b5f32a1e3b3a)
- [x] [add missing commas](https://github.com/pygments/pygments/pull/2259/commits/2ab3cf7c2f62506080f90c3e62013a5f8b7deca4)

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+                    Name.Constant)
+            ],
+            'character': [
+                (r'`.`', String.Char)
```
no escapes possible?

drkameleon replied:
Nope.

At least currently, character literals are always defined like this; that is a character enclosed in 2 backticks.

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+                        Name.Constant)
+            ],
+            'float': [
+                (r'[0-9]+\.[0-9]+',
```
both digit parts have to be present? (some languages allow `.5` or `2.`)

drkameleon replied:
@birkenfeld Hi! Original Arturo author here!

Floating-point numbers in arturo are always in the form of `(digits+)\.(digits+)` since `.(anything+)` is mostly used to identify `:literal` values: https://arturo-lang.io/documentation/language/#literal

P.S. In theory, it's only a `.` followed by any valid ID characters, but identifying `\.(digits+)` as something else (in that case a floating-point number) would make the whole thing rather confusing 😉

birkenfeld replied:
Thanks for clarifying!

RickBarretto replied:
Thank you, @drkameleon!

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+                (r'\:\w+',
+                        Keyword.Type)
+            ],
+            'attributes': [
```
this one is not used (another danger of using too many trivial states...)

RickBarretto replied:
- [x] [remove additional state](https://github.com/pygments/pygments/pull/2259/commits/ddb7923d1aa888f2bce3577bc44f744ad4f55c32)
  - I tried to just remove `constants` state, but I think the commit bellow is better
- [x] [remove the high amount of trivial states](https://github.com/pygments/pygments/pull/2259/commits/cb6a6b7595b15e8971794adc701e8eea1730198b)
  - I just removed that amount of states, put on `constants`, and added side notes. I think it's better.

birkenfeld replied:
There's still a few of them (punctuation, sugar, builtin-predicate-functions, maybe more...)

RickBarretto replied:
- [x] [remove trivial states](https://github.com/pygments/pygments/pull/2259/commits/fab78186a488bb081a4e8e0b6db7fc1866f3ff98)

RickBarretto replied:
@birkenfeld, do you think I need to remove `operators` state?

birkenfeld replied:
"need to" is too strong, but in general states that are only used as an include once are not technically needed.

RickBarretto replied:
Sure!

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,312 @@
+
+from pygments.util import ClassNotFound, get_bool_opt
+
+class ArturoLexer(RegexLexer):
```
please add a docstring with `versionadded`, look at other modules for examples

RickBarretto replied:
- [x] [add ArturoLexer's docstring](https://github.com/pygments/pygments/pull/2259/commits/fc625b63f15ce90e329c3523438a0cd63dec0473)

birkenfeld reviewed (changes requested):
Thanks for the patch!

A lot of comments, but nothing major.

RickBarretto commented:
Note: I'll test the file. Some of this changes may break something.

RickBarretto commented:
@birkenfeld. Can You help me with something?

So, in [these lines](https://github.com/pygments/pygments/pull/2259/files#diff-4d69c13e1da583622f39fd30da68656f10351065364fd699d9dd373db226cd8dR151-R166) I have string interpolation `string-interpol` and string templates `string-templates`. But, templates are working and interpol doesn't. 

They are almost the same, except for the notation. Templates start with `<||` and end with `||>`, while interpolation starts and ends with a `|` only.

---

Templates (working well):
![image](https://user-images.githubusercontent.com/78623871/197783771-09b70031-cbcc-435d-a4b7-9eee079f5e4f.png)

String Interpolation (broken):
![image](https://user-images.githubusercontent.com/78623871/197784106-2290a35f-ed06-4ddb-8b2d-9d0a64941729.png)

---
What should I do in this case?

birkenfeld commented:
In the string rules, you're matching `[^"]+` in the final rule, which slurps up the "|" that would 

The catchall rule must exclude all special characters, i.e. start of escape, start of interpolation and start of template.

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,283 @@
+    """
+    For Arturo source code
+
+    .. versionadded:: 0.9.80
```
should be the (next) Pygments version - 2.14.0

RickBarretto replied:
Ooh! Sorry, got it! My bad! 😂
I added current Arturo version.

RickBarretto replied:
- [x] [change versionadded](https://github.com/pygments/pygments/pull/2259/commits/67b45a9aa8a5a893ce9a7aa3edf163c49bf94b07)

RickBarretto commented:
> In the string rules, you're matching `[^"]+` in the final rule, which slurps up the "|" that would
> 
> The catchall rule must exclude all special characters, i.e. start of escape, start of interpolation and start of template.

- [x] [forget | when match String type](https://github.com/pygments/pygments/pull/2259/commits/1095047461d623e6c853f703561d92fc8137c54c)

RickBarretto commented:
- Issue: [W117: Overlap in character class: ['>']](https://github.com/pygments/pygments/actions/runs/3322527462/jobs/5492325895#step:5:9) 
  - [ ] I don't know how to fix it.
- Issue: [W119: Instead of negating character class, flip case of builtin class | step 5:12](https://github.com/pygments/pygments/actions/runs/3322527462/jobs/5492325895#step:5:12) and [step 5:15](https://github.com/pygments/pygments/actions/runs/3322527462/jobs/5492325895#step:5:15)
  - [x] fixed at [fix W119](https://github.com/pygments/pygments/pull/2259/commits/01d22d7c41737473d93b7d2f12a586134f3f6897)
- Issue: [Fail if mapfiles changed](https://github.com/pygments/pygments/actions/runs/3322527462/jobs/5492325729#step:5:12)
  - [x] fixed at [update mapfiles](https://github.com/pygments/pygments/pull/2259/commits/f938e5bf9b8d45456638030b022660d55ce13eed)

birkenfeld commented:
Re. W117: this usually happens when you have doubled entries in `words()`. And indeed, you have `'<:', ':>', ':<', ':>',` where I guess the last entry should be `>:`.

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,247 @@
+]
+class ArturoLexer(RegexLexer):
+    """
+    For Arturo source code
```
Could also add a link like in the PR description.

RickBarretto replied:
Sure!

RickBarretto commented:
> Thanks @RickBarretto for sticking with all the changes required and thanks Georg (@birkenfeld) for all the reviews and the assistance given. Good work! 👍

Thank you, @celtic-coder!!! 🚀

RickBarretto commented:
My bad, I forgot to update goldens. But I think it'll run now.

Error on my local machine gives me:
```diff
+ ' '           Text
- ' 2'          Literal.Number.Integer
?  -
+ '2'           Literal.Number.Integer
?    +...
```

So, It's right considering that a Integer must not have a spaces.

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,250 @@
+            include('string-escape'),
+            (r'\|', String.Interpol, 'inside-interpol'),        # Interpolation
+            (r'\<\|\|', String.Interpol, 'inside-template'),    # Templates
+            (r'\Z$', String.Single, '#pop'),    # Closing Quote
```
Isn't `\Z` end-of-string?

RickBarretto replied:
- [x] [remove redundant end-of-string](https://github.com/pygments/pygments/pull/2259/commits/e60c6c873b7aec67125389efe1b700374444beca)

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,250 @@
+    For Arturo source code
+
+    [Arturo's Github](https://github.com/arturo-lang/arturo)
+    [Arturo's Website](http://arturo-lang.io/)
```
This is reST markup, not Markdown...

RickBarretto replied:
- [x] [change hyperlinks to reST format.](https://github.com/pygments/pygments/pull/2259/commits/52ef9c821fcfb5925042374f62dd7b2e1d15dd78)

birkenfeld commented on tests/examplefiles/arturo/arturo_test.art.output:
```diff
@@ -0,0 +1,4999 @@
+'{'           Literal.String.Single
+'\n'          Literal.String
+
+' '           Literal.String
```
There's still a problem here with String contents getting matched char-by-char...

RickBarretto replied:
- [x] [replace generic Text with better tokens](https://github.com/pygments/pygments/pull/2259/commits/5ab2b42dd95f7511a10a8f328661cb08d9f05eb3)

RickBarretto replied:
It's happening only with multiline strings. I'll solve it!

RickBarretto replied:
How it's going now:
![image](https://user-images.githubusercontent.com/78623871/198035464-c3928f80-68be-490e-9337-b38164d208c7.png)

birkenfeld replied:
LGTM!

RickBarretto replied:
🙂

birkenfeld commented on pygments/lexers/arturo.py:
```diff
@@ -0,0 +1,250 @@
+            include('builtin-functions'),
+            include('operators'),
+
+            (r'.', Text),
```
There is a lot still being matched by this char-by-char...

RickBarretto commented:
> The \w metacharacter matches word characters. A word character is a character a-z, A-Z, 0-9, including _ (underscore).
> - [source](https://www.w3schools.com/JSREF/jsref_regexp_wordchar.asp)

My bad! I think it'll run now!

birkenfeld commented:
Added changelog and fixed up code style a bit in a followup commit.

celtic-coder commented:
Thanks Georg (@birkenfeld) for adding https://github.com/pygments/pygments/commit/8cb16e2bb67f28a84a703876ffdab335f53368f6.

RickBarretto commented:
Also, thank you @birkenfeld for your help, reviews, comprehension and patient. 🙂 
🎉 🚀

birkenfeld commented:
It was a pleasure :)
